### PR TITLE
exit program without deleting the source cluster if error occur

### PR DESCRIPTION
**Pull request by ca7alindev**

### CODE REVIEW

File: cmd/kndp/environment/move.go [hunk 1]
```diff
@@ -64,86 +63,95 @@ func (c *moveCmd) Run(ctx context.Context, logger *log.Logger) error {
...
+						Version:  paramsXRs.Spec.Versions[0].Name,
+						Resource: paramsXRs.Spec.Names.Plural,
+					}
+					for {
+						_, err = destClientset.Resource(resourceId).Namespace("").Create(ctx, &xr, metav1.CreateOptions{})
+						if err == nil {
+							logger.Info("Resource created successfully ", xr.GetName())
+							break
+						}
+						time.Sleep(5 * time.Second)
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:138]: @ca7alindev Hope you understand this is as hack, and can't be done like this in fail safe utilities like kndp.
Please make checks on configuration install level, where you have to check configuration health, and just after this continue with resources of configuration.

File: cmd/kndp/environment/move.go [hunk 2]
```diff
@@ -64,86 +63,95 @@ func (c *moveCmd) Run(ctx context.Context, logger *log.Logger) error {
...
+			})
+			if err != nil {
+				logger.Error(err)
+				return nil
+			}
+
+			for _, xr := range XRs {
+				xr.SetResourceVersion("")
+				labels := xr.GetLabels()
+				if labels != nil && labels["app.kubernetes.io/managed-by"] == "kndp" {
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:126]: @ca7alindev Add this to resources request as filter by label, this looks not good.
Just imagine you working with k8s API like with SQL DB, do you select all rows from DB and filter them, or add some condition as SQL WHERE? I think second, here shall be same.

**Review by web7evgheni — CHANGES_REQUESTED** [review]

File: cmd/kndp/environment/move.go [hunk 3]
```diff
@@ -21,8 +23,17 @@ type moveCmd struct {
...
+func checkHealthStatus(status []condition.Condition) bool {
+	healthStatus := false
+	for _, condition := range status {
+		if condition.Type == "Healthy" && condition.Status == "True" {
+			healthStatus = true
+		}
+	}
+	return healthStatus
+}
+
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:35]: @ca7alindev to internal

File: cmd/kndp/environment/move.go [hunk 4]
```diff
@@ -64,86 +76,124 @@ func (c *moveCmd) Run(ctx context.Context, logger *log.Logger) error {
...
+			for _, conf := range destConf {
+				var paramsConf configuration.Configuration
+				if err := runtime.DefaultUnstructuredConverter.FromUnstructured(conf.UnstructuredContent(), &paramsConf); err != nil {
+					logger.Printf("Failed to convert item %s: %v\n", conf.GetName(), err)
+					continue
+				}
+				status := paramsConf.Status.Conditions
+				healthStatus := checkHealthStatus(status)
+				if healthStatus {
+					outerLoopBreak = true
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:123]: @ca7alindev Here you check just fist configuration health, next configs are not verified, please be more attentive.

File: cmd/kndp/environment/move.go [hunk 5]
```diff
@@ -64,86 +76,124 @@ func (c *moveCmd) Run(ctx context.Context, logger *log.Logger) error {
...
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(xrd.UnstructuredContent(), &paramsXRs); err != nil {
+				logger.Printf("Failed to convert item %s: %v\n", xrd.GetName(), err)
+				return nil
+			}
+
+			XRs, err := kube.GetKubeResources(kube.ResourceParams{
+				Dynamic:   sourceDynamicClient,
+				Ctx:       ctx,
+				Group:     paramsXRs.Spec.Group,
+				Version:   paramsXRs.Spec.Versions[0].Name,
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:160]: @ca7alindev All versions please.

File: cmd/kndp/environment/move.go [hunk 6]
```diff
@@ -21,8 +23,17 @@ type moveCmd struct {
...
+	healthStatus := false
+	for _, condition := range status {
+		if condition.Type == "Healthy" && condition.Status == "True" {
+			healthStatus = true
+		}
+	}
+	return healthStatus
+}
+
 func (c *moveCmd) Run(ctx context.Context, logger *log.Logger) error {
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:36]: @ca7alindev Please move all logic to internal package, in command remains just parameters parsing if is required.
Main cause for that is later usage by TUI and other commands.

**Review by web7evgheni — CHANGES_REQUESTED** [review]

File: internal/environment/move.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+		return nil, err
+	}
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	return dynamicClient, nil
+}
```
Comment (web7evgheni) [inline internal/environment/move.go:44]: @ca7alindev move this to kube package

File: cmd/kndp/environment/move.go [hunk 8]
```diff
@@ -22,128 +19,64 @@ type moveCmd struct {
...
-		Ctx:       ctx,
-		Group:     "apiextensions.crossplane.io",
-		Version:   "v1",
-		Resource:  "compositeresourcedefinitions",
-		Namespace: "",
+	//Delete source cluster after all resources are successfully created in destination cluster
+	if err == nil {
+		cmd := exec.Command("kind", "delete", "cluster", "--name", strings.TrimPrefix(c.Source, "kind-"))
+		cmd.Run()
+		logger.Info("Successfully moved Kubernetes resources to the destination cluster.")
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:78]: @ca7alindev move all of this to environment package.

**Review by web7evgheni — CHANGES_REQUESTED** [review]

File: cmd/kndp/environment/move.go (hunk 8 above)
Comment (ca7alindev) [reply cmd/kndp/environment/move.go:78]: @web7evgheni, but this logic is already in environment package under cmd/kndp/environment

File: internal/kube/client.go [hunk 9]
```diff
@@ -40,6 +45,19 @@ const (
 	UpboundK8sResource = "k8s"
 )
 
+func CreateKubernetesClients(ctx context.Context, logger *logger.Logger, context string) *dynamic.DynamicClient {
```
Comment (web7evgheni) [inline internal/kube/client.go:48]: @ca7alindev According to Go naming convention, please rename it to Context()
https://google.github.io/styleguide/go/best-practices#naming-conventions

File: cmd/kndp/environment/move.go [hunk 10]
```diff
@@ -22,128 +20,56 @@ type moveCmd struct {
...
```
Comment (web7evgheni) [inline cmd/kndp/environment/move.go:71]: @ca7alindev Please collect all this sources inside of single function in internal/environment/move.go and call it in command function, how we discussed, all model related sources shall be outside of controller.

**Review by web7evgheni — CHANGES_REQUESTED** [review]

**Review by web7evgheni — APPROVED** [review]

**Pull request merged**